File: solutions/020/020.js
/**
 * Project Euler - Problem Solution 020
 * Copyright (c) Justin McGettigan. All rights reserved.
 * https://github.com/jwmcgettigan/project-euler-solutions
 */

const factorial = (n) => {
  // reused from problem 015 w/BigInt added
  if(n < 2) { return BigInt(1); }
  else { return BigInt(n) * factorial(n-1); }
};

const sumDigits = (n) => {
  // reused from problem 016
  let s = BigInt(0), base = BigInt(10);
  while(n > 0) {
    remainder = n % base;
    s += remainder;
    n = (n - remainder) / base;
  }
  return Number(s);
};

const factorialDigitSum = (n) => {
  return sumDigits(factorial(n));
};

console.log(factorialDigitSum(100));

File: solutions/020/020.py
"""
Project Euler - Problem Solution 020
Copyright (c) Justin McGettigan. All rights reserved.
https://github.com/jwmcgettigan/project-euler-solutions
"""

def factorial(n):
  # reused from problem 015
  if n < 2: return 1
  else: return n * factorial(n-1)

def sum_digits(n):
  # reused from problem 016
  s = 0
  while n:
    s, n = s + n % 10, n // 10
  return s

def factorial_digit_sum(n):
  return sum_digits(factorial(n))
  

if __name__ == "__main__":
  print(factorial_digit_sum(100))